File: mcp-code-intelligence-nodejs/src/analyzers/complexity/ComplexityCalculator.ts
/**
 * KSA-161: Core complexity calculation engine.
 * Delegates to language-specific counters via strategy pattern.
 */

import type { SyntaxNode } from '../../parsers/types.js';
import type { ComplexityBreakdown } from './types.js';
import { BaseNodeCounter } from './counters/BaseNodeCounter.js';
import { TypeScriptCounter } from './counters/TypeScriptCounter.js';
import { PythonCounter } from './counters/PythonCounter.js';
import { JavaCounter } from './counters/JavaCounter.js';
import { KotlinCounter } from './counters/KotlinCounter.js';
import { GoCounter } from './counters/GoCounter.js';

export class ComplexityCalculator {
  private counters: Map<string, BaseNodeCounter> = new Map();

  constructor() {
    // Register built-in counters
    this.registerCounter(new TypeScriptCounter());
    this.registerCounter(new PythonCounter());
    this.registerCounter(new JavaCounter());
    this.registerCounter(new KotlinCounter());
    this.registerCounter(new GoCounter());
    // JS uses same rules as TS
    this.counters.set('javascript', new TypeScriptCounter());
  }

  /** Register a language-specific counter. */
  registerCounter(counter: BaseNodeCounter): void {
    this.counters.set(counter.language, counter);
  }

  /** Calculate cyclomatic complexity for a function body AST node. */
  calculate(bodyNode: SyntaxNode, language: string): ComplexityBreakdown | null {
    const counter = this.counters.get(language);
    if (!counter) return null;

    const counts = counter.countDecisionPoints(bodyNode);
    const nestingDepth = counter.calculateNestingDepth(bodyNode);
    const earlyReturns = counter.countEarlyReturns(bodyNode);

    const cc = 1 + counts.branches + counts.loops + counts.logical_ops + counts.exception_handlers;

    return {
      cyclomatic_complexity: cc,
      branches: counts.branches,
      loops: counts.loops,
      logical_ops: counts.logical_ops,
      exception_handlers: counts.exception_handlers,
      nesting_depth: nestingDepth,
      early_returns: earlyReturns,
    };
  }

  /** Check if a language is supported. */
  supportsLanguage(language: string): boolean {
    return this.counters.has(language);
  }

  /** Get list of supported languages. */
  getSupportedLanguages(): string[] {
    return [...this.counters.keys()];
  }
}

File: backend/mcp-code-intelligence-kotlin/src/main/kotlin/com/codeintel/analyzers/complexity/ComplexityCalculator.kt
/**
 * KSA-161: Core complexity calculation engine.
 * Delegates to language-specific counters via strategy pattern.
 */
package com.codeintel.analyzers.complexity

import com.codeintel.analyzers.complexity.counters.*
import com.codeintel.parsers.SyntaxNode

class ComplexityCalculator {
    private val counters = mutableMapOf<String, BaseNodeCounter>()

    init {
        registerCounter(TypeScriptCounter())
        registerCounter(PythonCounter())
        registerCounter(JavaCounter())
        registerCounter(KotlinCounter())
        registerCounter(GoCounter())
        counters["javascript"] = TypeScriptCounter()
    }

    fun registerCounter(counter: BaseNodeCounter) {
        counters[counter.language] = counter
    }

    fun calculate(bodyNode: SyntaxNode, language: String): ComplexityBreakdown? {
        val counter = counters[language] ?: return null
        val counts = counter.countDecisionPoints(bodyNode)
        val nestingDepth = counter.calculateNestingDepth(bodyNode)
        val earlyReturns = counter.countEarlyReturns(bodyNode)
        val cc = 1 + counts.branches + counts.loops +
            counts.logicalOps + counts.exceptionHandlers
        return ComplexityBreakdown(
            cyclomaticComplexity = cc,
            branches = counts.branches,
            loops = counts.loops,
            logicalOps = counts.logicalOps,
            exceptionHandlers = counts.exceptionHandlers,
            nestingDepth = nestingDepth,
            earlyReturns = earlyReturns,
        )
    }

    fun supportsLanguage(language: String): Boolean =
        counters.containsKey(language)

    fun getSupportedLanguages(): List<String> = counters.keys.toList()
}
